### ISSUE TITLE
Improve user experience of 10-10CG related to long submission times

### ISSUE THREAD

**Issue by mchelen-gov** [department-of-veterans-affairs/va.gov-team#19321]
As a Veteran submitting a 10-10CG application
I would like to be better informed about what is happening when the submission is taking a long time to go through
So that I can be confident in my submission and the relevant next steps.

AC
- [x] Investigate and determine root cause of long submission times
- [x] Veterans informed that submissions may take a long time
- [x] Veterans see a visual indicator that the submission is working as expected (even if it's taking a long time)
- [ ] ~Veterans understand submission status and next steps after a minute of waiting~ (This work will be done in #21220)
- [x] Users are pointed to the error message that informs them that the submission did not go through

Ancillary work:
- Platform: [Troubleshoot timeout issues for VSA Caregiver connection to Salesforce/CARMA](https://github.com/department-of-veterans-affairs/va.gov-team/issues/19503)

Reference:
- https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/teams/vsa/teams/caregiver/architecture-diagram-as-of-june-15-2020.pdf

**Comment by ala-yna** [comment]
![Screen Shot 2021-02-18 at 10.23.25 AM.png](https://images.zenhubusercontent.com/5ddc08a57dcda00001c763e1/151ff8eb-992c-46fa-b3b3-8b68ccbb0f98)

**Comment by ala-yna** [comment]
Count of timeout errors by date since 10/1:

| Date (UTC)  | COUNTA of Date (UTC) |
|-------------|----------------------|
|  10/29/2020 |                    1 |
|   11/3/2020 |                    1 |
|   11/5/2020 |                    1 |
|   12/8/2020 |                    1 |
|  12/22/2020 |                    2 |
|    1/4/2021 |                    1 |
|    1/6/2021 |                    2 |
|    1/7/2021 |                    1 |
|    1/8/2021 |                    1 |
|    1/9/2021 |                    2 |
|   1/10/2021 |                    1 |
|   1/11/2021 |                    1 |
|   1/13/2021 |                    3 |
|   1/14/2021 |                    2 |
|   1/15/2021 |                    1 |
|   1/16/2021 |                    2 |
|   1/17/2021 |                    2 |
|   1/20/2021 |                    4 |
|   1/21/2021 |                    1 |
|   1/22/2021 |                    3 |
|   1/23/2021 |                    2 |
|   1/24/2021 |                    4 |
|   1/25/2021 |                    3 |
|   1/26/2021 |                    5 |
|   1/27/2021 |                    1 |
|   1/28/2021 |                   29 |
|   1/29/2021 |                    7 |
|   1/30/2021 |                    5 |
|   1/31/2021 |                    6 |
|    2/1/2021 |                    2 |
|    2/2/2021 |                    4 |
|    2/3/2021 |                   22 |
|    2/4/2021 |                    8 |
|    2/5/2021 |                    4 |
|    2/6/2021 |                    5 |
|    2/7/2021 |                    3 |
|    2/8/2021 |                    2 |
|    2/9/2021 |                    3 |
|   2/10/2021 |                    1 |
|   2/11/2021 |                    9 |
|   2/12/2021 |                    2 |
|   2/13/2021 |                    6 |
|   2/15/2021 |                    3 |
|   2/16/2021 |                    3 |
|   2/17/2021 |                    4 |
| Grand Total |                  176 |


![Screen Shot 2021-02-19 at 12.53.01 PM.png](https://images.zenhubusercontent.com/5ddc08a57dcda00001c763e1/03daf7fd-be4a-477e-929b-8b83ba21fc49)

**Comment by mchelen-gov** [comment]
Additional metrics reference: https://docs.google.com/spreadsheets/d/1OVghvx8yFkOSNnnzpHZulbdDnqQRGohAOr4SELO0L7M

**Comment by kevinmirc** [comment]
Wanted to document some of the solutions we proposed to the Caregiver Progam in our last sync (Friday, February 19, 2021). These are some ideas that we can consider to make the user's experience a little bit better when processing time is long (knowing it's 10 avg and up to 60+ seconds).

- Set expectations on processing time before submission.
- Provide progressive feedback when data is being processed.
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/19505 (when a timeout occurs) Inform users that the result of their submission is **unknown** and providing the next steps: "Contact {processing facility selected by the user} or CSL".
- Search CARMA by the claim ID (will allow us to get status when the CARMA Case ID is not returned to VA.gov).

<img width="1396" alt="Screen Shot 2021-02-22 at 9 25 46 AM" src="https://user-images.githubusercontent.com/7391284/108721766-61fee580-74f0-11eb-94dd-10c1cc71d647.png">

<img width="1396" alt="Screen Shot 2021-02-22 at 9 25 40 AM" src="https://user-images.githubusercontent.com/7391284/108721762-61664f00-74f0-11eb-8b40-d113405f9f68.png">

<img width="1397" alt="Screen Shot 2021-02-22 at 9 32 31 AM" src="https://user-images.githubusercontent.com/7391284/108722235-e8b3c280-74f0-11eb-90f5-e01cda276477.png">

**Comment by mchelen-gov** [comment]
for reference: https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/teams/vsa/teams/caregiver/architecture-diagram-as-of-june-15-2020.pdf

**Comment by ala-yna** [comment]
# Notes from 2/18

Investigation as a part of #19321 revealed that increasing the timeout to 2 minutes will resolve about half of the timeout errors.

Conclusion is that some users will have to wait 60-120 seconds to see that their application went through. There is currently no visible way to see that the submission process is "working" during this window. How can we give users a better sense that the submission process is working as expected?

Constraints:

Some applications are rejected for valid reasons by CARMA after processing (duplicate submission, etc)
Next steps:
1) Reach out to CARMA to understand how they plan to address long data processing times
2) Understand how other teams are addressing extended data processing times
3) Check in with design system on if there's a preferred "waiting room" type UX
4) Design spike to generate solutions

**Comment by ala-yna** [comment]
# Notes from 2/22 
**Immediate next steps:**
- [x] Consult with design system team

**Near term solution:** (keep timeout to 60 seconds)
- Info (prior to submission) that it may take up to a minute
- Visual indicator (spinner + progressive messaging) 
- Timeout message after a minute (+ facility to call?)

**Midterm solution**
- Explore how to change BE/UI to accommodate longer submission times
- Give users a way to check in on (just submission) status

**Longer term solution**
- Explore giving Veterans a way to check on the status of their application

**Comment by ala-yna** [comment]
# Team discussion on 2/24

- [ ] Design: Generate content for progressive spinner
- [ ] Frontend: Create progressive spinner component
- [ ] Frontend: Confirm JS timeout is set to 120 seconds
- [ ] Design: Generate content/mockup to set expectation around time
- [ ] Frontend: Add progressive loading message to online 10-10CG

**Comment by kevinmirc** [comment]
**Summary** of the state of this epic at this time: 
- The CARMA forward proxy was extended for production [#8582](https://github.com/department-of-veterans-affairs/devops/pull/8582) and staging [#8574](https://github.com/department-of-veterans-affairs/devops/pull/8574).
- Extending the timeout of the submission request in vets-api [#5906](https://github.com/department-of-veterans-affairs/vets-api/pull/5900), is still on hold (based on revproxy needing changes - see below).
- ⚠️ When testing the progressive loading indicator [#16166](https://github.com/department-of-veterans-affairs/vets-website/pull/16166), we discover that the reverse proxy will timeout out at exactly 60 seconds. This means:
  - The revproxy is responding to the request before our API does
    - making the res status in our longs for 60s+ submissions, not the res status the user experiences.
  - Users are getting a completely different user experience that what we assumed (see [recording](https://dsva.slack.com/archives/CMJ2V70UV/p1614359494043800))
    - The are getting a generic network error message (see attached).
    - They **can** click the submit button again (see attached).
    - A global error handler is overriding our custom error ([details here](https://dsva.slack.com/archives/CBU0KDSB1/p1614701492468100?thread_ts=1614627021.416200&cid=CBU0KDSB1)).
  - The revproxy would need to be extended for the entire request duration (not just carma submission).
- [Looked into GA](https://dsva.slack.com/archives/CMJ2V70UV/p1614696974062900) and confirmed that a user who experienced a timeout error was able to apply again, within 4 minutes, without any additional navigation.


**Images**:
- After revproxy timeout occurs
  - the user can press submit again:
    - <img width="500" alt="Screen Shot 2021-03-03 at 2 43 35 PM" src="https://user-images.githubusercontent.com/7391284/109862822-e3602180-7c2e-11eb-8759-58fb1b2f2a57.png">
  - a generic error message is shown:
    - <img width="500" alt="Screen Shot 2021-03-03 at 2 43 45 PM" src="https://user-images.githubusercontent.com/7391284/109862832-e5c27b80-7c2e-11eb-9d1e-f4f4042f120b.png">

**Next Steps**:
- Will be looking into more user sessions to see when and how users re-submit [#20741](https://github.com/department-of-veterans-affairs/va.gov-team/issues/20741)
- Extending the timeout as a short term solution will need to be re-evaluated (blocked by the step below)
  - Because configuration needs to be changed in multiple places, and those changes are based on external system's performce, it will over complicate our system. That, along with the fact that users appear to be able to easily reapply, this short term "solution" may be worth skipping for a better, more appropriate resolution.
- Gain an understanding from the program's preference of how to handle timed-out submissions
  - When a timeout occurred, do they prefer to have a duplicate submission with the attachment and the veteran is assured that it went through, or have the veteran contact someone to confirm that it went through before re-applying.

**Comment by kevinmirc** [comment]
@mlighdc and I had a discussion today about allowing the user to retry submission when CARMA times out.

Doing so would of course increase the chance/number of duplicates. We can add a limit for retries (one reattempt perhaps) but based on previous research (_insert ticket number_), of the 10 submissions we looked at, less than 50% of resubmissions actually went through - so this is not a great solution.

Before moving forward with providing users with a better retry experience on timeout, @mlighdc suggested we connect with Ranga and the CARMA team to explore these options:

1. CARMA sending vets-api an acknowledgment within 60 seconds (even if it's still processing)
2. Making duplicate submissions easier for CSCs to identify
3. Rolling back the transaction if it takes longer than _x_ seconds

The last option would us to allow us to (1) accurately tell the user the application did not go through, and (2) enable to user to retry without the likely risk of duplicates.

The first option has the negative consequence of CARMA accepting an application before knowing that it can be created as a record.
